### PR TITLE
Always use 'named' exports in rollup.config.js

### PR DESCRIPTION
**Pull request by Andarist**
**What**:

Always use 'named' exports in rollup.config.js

**Why**:

Differentiating them was imho a mistake and also a breaking change when you have switched from using babel to rollup. It is also a reason why you need to do [such hacks](https://github.com/paypal/downshift/blob/master/src/index.js#L4-L9) in possibly every library using kcd-scripts. Using mentioned hack is preventing tree-shaking to take place, which is also a big downside of this approach. (I understand that i.e. downshift is "used or not" and it doesnt have to be any more tree-shakeable than that, but consider downshift being a dependency of other package - if one imports other package which in turns uses downshift as part of its tree-shakeable named export - downshift might not get tree-shaken from the bundle **at all**).

(!) Unfortunately this is gonna be a breaking change.

**Checklist**:

* [ ] Documentation N/A
* [ ] Tests N/A
* [ ] Ready to be merged >
* [ ] Added myself to contributors table

### CODE REVIEW

**Comment by kentcdodds** [comment]
Hmmm... I'll have to give this some thought. I'm not certain that I understand the implications. I know that I looked into it thoroughly when I created the config originally and I thought I had it right so it would tree-shake properly. I thought the hacks were for TypeScript weirdness only...

Anyway. I'll try to get to this as soon as I can. Thanks!

**Comment by Andarist** [comment]
From what I understand TS most commonly use "node" resolution algorithm. This means that when resolving a package it will reach out to "main" field in package.json exclusively - ignoring "module" - which is in a commonjs format.

If your package exports only a default export rollup will produce
```js
module.exports = Downshift$1
```
by default for better interop with node. Problem is that now TS compiler wont be able to tell that when you do
```js
import Downshift from 'downshift'
```
it wont be able to tell that your module has `default` export. How would it? By adding this "hack" to your codebase it transpiles down to
```js
Downshift$1.default = Downshift$1;
module.exports = Downshift$1;
```
Which allows TS to resolve default export.

But now we end with a weirdo situation, where
```js
require('downshift').default === require('downshift')
```

Not mentioning that whenever your library adds additional named export it makes it automatically a breaking change for that library (regarding cjs consumers) and it's really easy to miss that, because you do not inspect cjs output after each build (who does? 😉) - you only care about what you write to `src` and even your tests might be transpiled with `babel-jest` or something and you won't notice that breaking change until releasing it to the audience.

**Regarding tree-shakeability**
webpack is not actually doing full "tree-shaking" it just removes assignments to module.exports and let uglifyJS to do dead code elimination later and unfortunately static properties (I advise against using them without good reason) prevents DCE - this TS hack is a static property assignment, so let's examine simple scenarios

Input:
```js
var Test = function () {};
```
Output from uglifyJS
```js
```

Input:
```js
var Test = function () {};
Test.default = test;
```
Output from uglifyJS
```js
(function(){}).default=test;
```

Note - I've used such command on the sample file `uglifyjs -mc passes=3,unsafe --toplevel`

**Comment by kentcdodds** [comment]
Thanks for explaining. I was tired last night 😪 After a night's rest this makes much more sense.

So here's the thing I was trying to avoid by having it different depending on the format. I didn't want people using CJS or globals to have to do `require('downshift').default` or `window.Downshift.default` because that's super annoying.

That said, I think I later fixed _that_ as well by adding the different `-entry` support ([like this](https://github.com/paypal/glamorous/blob/master/src/cjs-entry.js))...

I'll have to do a little testing around this. Thanks for bringing it up!

**Comment by Andarist** [comment]
> So here's the thing I was trying to avoid by having it different depending on the format. I didn't want people using CJS or globals to have to do require('downshift').default or window.Downshift.default because that's super annoying.

Yeah, I understand that - many people dont like `.default` and add things like [`babel-plugin-add-module-exports`](https://github.com/59naga/babel-plugin-add-module-exports) to their configs. I dont consider it bad in general, but I think this is kinda a dangerous practice, because it's so easy to introduce a breaking change without realizing it. Would be cool to find a way to test for such accidental breaking changes and to warn user about it, but dont have idea at the moment how to do that automatically (beside caching info about previous builds somewhere).

> That said, I think I later fixed that as well by adding the different -entry support (like this)...

Interesting, never though about creating a different entry designated for a otherwise normal cjs entry. Does your kcd-scripts package recognize this with a string pattern and build cjs using such entries? Or do you have to configure it when running a build script (dont see it in the package.json of glamorous so I'd expect the the former)?

**Comment by kentcdodds** [comment]
Yeah, `kcd-scripts build` will create 4 files via rollup. One for each format. The `input` is either `src/{format}-entry.js` if it exists or `src/index.js` if it doesn't. Kinda handy for stuff like that.

But I guess the fact that TypeScript reads the `main` when you import modules messes everything up. Ugh... Are you aware of whether Flow has the same problem?

**Comment by Andarist** [comment]
> Are you aware of whether Flow has the same problem?

Not sure, but can test this later. 

> Yeah, kcd-scripts build will create 4 files via rollup. One for each format. The input is either src/{format}-entry.js if it exists or src/index.js if it doesn't. Kinda handy for stuff like that.

Interesting, gonna think about it some more, but the idea seems promising for my projects.

**Comment by kentcdodds** [comment]
Hmmm... Would we be able to remove the TypeScript hack if I used this for the CJS entry file?

```javascript
/* istanbul ignore next */
import * as libStar from './'

const lib = libStar.default

Object.assign(
  lib,
  Object.keys(libStar).reduce((e, prop) => {
    if (prop !== 'default') {
      // eslint-disable-next-line import/namespace
      e[prop] = libStar[prop]
    }
    return e
  }, {}),
)

lib.default = lib

export default lib
```

This way TypeScript is happy because the `default` property is the `lib` so when people `import defaultThing from 'thing'` TypeScript will be able to resolve that right? And then the ESM part will still be tree-shakeable.

I think this should work. What do you think?

**Comment by Andarist** [comment]
> This way TypeScript is happy because the default property is the lib so when people import defaultThing from 'thing' TypeScript will be able to resolve that right? And then the ESM part will still be tree-shakeable.

I think it should, although im not a TS user and i dont know if it will understand creating those named exports dynamically like this - I guess it should as its technically statically analyzable. 

> And then the ESM part will still be tree-shakeable.

Problem will still be there, at least for `downshift`, because it ["exports"](https://github.com/paypal/downshift/blob/20538af4a979fbfbc4686bf919e7f08f5b78add7/src/index.js#L10) also a static `resetIdCounter`, whereas it should be imho a named export.

**Comment by kentcdodds** [comment]
> i dont know if it will understand creating those named exports dynamically like this - I guess it should as its technically statically analyzable.

Perhaps we don't have to do it in a loop and can just do things manually. Either that or we could use codegen (we're doing that right now in glamorous anyway: see the [code](https://github.com/paypal/glamorous/blob/b564686491ede7709435e94de10a61dfe743b1ec/src/index.js#L64-L70) and look at the bottom of [this file](https://unpkg.com/glamorous@4.11.3/dist/glamorous.esm.js) for the result.

And in downshift I would change that to a named export. I don't like it as a property the `Downshift` component anyway. But in either case, tree-shaking that function is hardly necessary, so no big deal :) I'm thinking more about situations where tree-shaking would really be useful :)

I think I'll try implementing this in glamorous and downshift and we'll see if it works out alright.

**Comment by Andarist** [comment]
> And in downshift I would change that to a named export. I don't like it as a property the Downshift component anyway. But in either case, tree-shaking that function is hardly necessary, so no big deal :) I'm thinking more about situations where tree-shaking would really be useful :)

There is a possibility (actual one, not in the sense that it im not sure) that such static property can prevent tree-shaking whole downshift if downshift is used as dependency of some project and it gets imported but not used. Having this simple static utility tree-shakable would be an additional benefit ofc, but a really small one.

It is a little bit hard to explain, but es modules are eager if you put an import statement in your code it gets hoisted and executed right away which ofc can cause a cascade of imports, because your dep might rely on other things. Having this in mind a module can have a side effect and tools are "afraid" of dropping potentially effectful statements. Such static assignment is treated by most tools (uglifyjs for sure) as unsafe to remove.

So if we have a project A which depends on B which depends on downshift, all that code have to be included in the final "memory" bundle, once we have a dep graph we can start tree-shaking, but we can only tree-shake if we are sure its safe to do so. So even if we recognize that code inside B can be removed it in fact has "required" downshift to be included and tools have hard time to determine if downshift can be removed.

That being said it currently cannot be removed anyway because the component uses static properties anyway - but babel@7 has a pending PR to include static properties in "class IIFE" (it is an IIFE generated by class transforms, you'd probably see it when inspecting classes transpiled down to es5) which gets also annotated with "magic" `#__PURE__` comments which are supposed to tell tools like uglifyJS that this particular IIFE is pure and if its result stays unused it can be removed entirely.

To sum it up - moving this thing to named export would be a step towards making the situation better.

**Comment by kentcdodds** [comment]
That makes sense. Thanks for taking time to explain. I think it's about time I work on downshift v2 anyway :)

**Comment by kentcdodds** [comment]
In any case, I don't believe that this PR's changes will be necessary, so I'll go ahead and close it (though you can feel free to continue the conversation if you like).

**Comment by Andarist** [comment]
The only problem that I see with your proposed approach is that it needs to be implemented in each library, whereas just using more interoperable exports in the tool would solve this once for all.

**Comment by kentcdodds** [comment]
True. I really want to avoid `require('foo').default` though. Do you think I could accomplish both?

**Comment by Andarist** [comment]
**Maybe** it would be possible with custom rollup plugin that would create your `cjs-entry.js` on the fly in memory. Dunno if you can override main entry point with `resolveId` hook though, u'd have to experiment a little bit.
